### ISSUE TITLE
[BUG]

### ISSUE THREAD

**Issue by colindawson** [docker-mailserver/docker-mailserver#2113]
# Bug Report

Clam AV is failing when running in kubernetes

Aug  7 20:40:12 mailserver-676694cbcb-9ln82 amavis[544]: (00544-01) (!)run_av (ClamAV-clamd) FAILED - unexpected , output="/var/lib/amavis/tmp/amavis-20210807T204011-00544-Uzzrwj3z/parts: lstat() failed: Permission denied. ERROR\n"
Aug  7 20:40:12 mailserver-676694cbcb-9ln82 amavis[544]: (00544-01) (!)ClamAV-clamd av-scanner FAILED: CODE(0xaaaadd3dab40) unexpected , output="/var/lib/amavis/tmp/amavis-20210807T204011-00544-Uzzrwj3z/parts: lstat() failed: Permission denied. ERROR\n" at (eval 98) line 951.
Aug  7 20:40:12 mailserver-676694cbcb-9ln82 amavis[544]: (00544-01) (!)WARN: all primary virus scanners failed, considering backups
Aug  7 20:40:25 mailserver-676694cbcb-9ln82 postfix/smtpd[1668]: timeout after END-OF-MESSAGE from localhost[127.0.0.1]

## Context

I have no volumes mounted for the clamAV folders, as I didn't see anything in the docker-hub info, in the documentation wiki or in any open issues.

### What is affected by this bug?

it looks like ClamAV isn't working, although the email sources are reporting that they've been scanned!

### When does this occur?

Every time an e-mail is received.

### How do we replicate the issue?

1. Start docker image  mailserver/docker-mailserver:10.0.0
2. Recieve an email.
3.
4.

### Actual Behavior

Aug  7 20:46:42 mailserver-676694cbcb-9ln82 postfix/smtpd[4874]: disconnect from mail-oi1-f178.google.com[209.85.167.178] ehlo=2 starttls=1 mail=1 rcpt=1 bdat=1 quit=1 commands=7
Aug  7 20:46:42 mailserver-676694cbcb-9ln82 amavis[543]: (00543-02) (!)run_av (ClamAV-clamd) FAILED - unexpected , output="/var/lib/amavis/tmp/amavis-20210807T203421-00543-tG0afhAw/parts: lstat() failed: Permission denied. ERROR\n"
Aug  7 20:46:42 mailserver-676694cbcb-9ln82 amavis[543]: (00543-02) (!)ClamAV-clamd av-scanner FAILED: CODE(0xaaaadd3dab40) unexpected , output="/var/lib/amavis/tmp/amavis-20210807T203421-00543-tG0afhAw/parts: lstat() failed: Permission denied. ERROR\n" at (eval 98) line 951.
Aug  7 20:46:42 mailserver-676694cbcb-9ln82 amavis[543]: (00543-02) (!)WARN: all primary virus scanners failed, considering backups

### Expected Behavior

There should not be a permission error.

## Your Environment

- version: `10.0.0`
- available RAM: `8GB`

- Ubuntu 20.04.2 LTS
- docker://20.10.7
- v1.21.3

### Environment Variables

```
kind: ConfigMap
apiVersion: v1
metadata:
  name: mailserver.env.config
  namespace: mailserver
  labels:
    app: mailserver
data:
  POSTFIX_MESSAGE_SIZE_LIMIT: "100000000"
  OVERRIDE_HOSTNAME: <guess why I've changed this>
  ENABLE_FETCHMAIL: "0"
  FETCHMAIL_POLL: "120"
  ENABLE_SPAMASSASSIN: "1"
  SPAMASSASSIN_SPAM_TO_INBOX: "1"
  SA_SPAM_SUBJECT: "undef"
  MOVE_SPAM_TO_JUNK: "1"
  SA_TAG: "-100000.0"
  SA_TAG2: "4"
  SA_KILL: "100000.0"
  ENABLE_AMAVIS: "1"
  AMAVIS_LOGLEVEL: "0"
  ENABLE_CLAMAV: "1"
  ENABLE_FAIL2BAN: "1"
  ENABLE_POSTGREY: "1"
  ONE_DIR: "1"
  DMS_DEBUG: "0"
  SSL_TYPE: "manual"
  SSL_CERT_PATH: "/etc/ssl/mailserver/tls.crt"
  SSL_KEY_PATH: "/etc/ssl/mailserver/tls.key"

apiVersion: apps/v1
kind: Deployment
metadata:
  name: mailserver
  namespace: mailserver
spec:
  replicas: 1
  selector:
    matchLabels:
      app: mailserver
  template:
    metadata:
      labels:
        app: mailserver
        role: mail
        tier: backend
    spec:
      #nodeSelector:
      #  kubernetes.io/hostname: local.k8s
      #initContainers:
      #- name: init-myservice
      #  image: busybox
      #  command: ["/bin/sh", "-c", "cp /tmp/user-patches.sh /tmp/files"]
      #  volumeMounts:
      #    - name: config
      #      subPath: user-patches.sh
      #      mountPath: /tmp/user-patches.sh
      #      readOnly: true
      #    - name: tmp-files
      #      mountPath: /tmp/files
      containers:
      - name: docker-mailserver
        image: mailserver/docker-mailserver:10.0.0
        imagePullPolicy: Always
        securityContext:
          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: false
          runAsUser: 0
          runAsGroup: 0
          runAsNonRoot: false
          privileged: false
          capabilities:
            add:
              # file permission capabilities
              - CHOWN
              - FOWNER
              - MKNOD
              - SETGID
              - SETUID
              - DAC_OVERRIDE
              # miscellaneous  capabilities
              - SYS_CHROOT
              - NET_BIND_SERVICE
              - KILL
              - NET_ADMIN
            drop: [ALL]
          seccompProfile:
            type: RuntimeDefault
        volumeMounts:
          - name: postfix-accounts
            mountPath: /tmp/docker-mailserver/postfix-accounts.cf
            readOnly: true
          - name: config
            subPath: postfix-main.cf
            mountPath: /tmp/docker-mailserver/postfix-main.cf
            readOnly: true
          - name: config
            subPath: postfix-virtual.cf
            mountPath: /tmp/docker-mailserver/postfix-virtual.cf
            readOnly: true
          - name: config
            subPath: fetchmail.cf
            mountPath: /tmp/docker-mailserver/fetchmail.cf
            readOnly: true
          - name: config
            subPath: dovecot.cf
            mountPath: /tmp/docker-mailserver/dovecot.cf
            readOnly: true
          - name: opendkim
            mountPath: /tmp/docker-mailserver/opendkim
            readOnly: true
          - name: data
            mountPath: /var/mail
            subPath: data
          - name: data
            mountPath: /var/mail-state
            subPath: state
          - name: data
            mountPath: /var/log/mail
            subPath: log
          - name: tls
            mountPath: /etc/ssl/mailserver
            readOnly: true
        ports:
          - name: smtp
            containerPort: 25
            protocol: TCP
          - name: smtp-secure
            containerPort: 465
            protocol: TCP
          - name: smtp-auth
            containerPort: 587
          - name: imap
            containerPort: 143
            protocol: TCP
          - name: imap-secure
            containerPort: 993
            protocol: TCP
        envFrom:
          - configMapRef:
              name: mailserver.env.config
      volumes:
        - name: config
          configMap:
            name: mailserver.config
        - name: postfix-accounts
          hostPath:
             path: /mnt/k8s/yaml/mailserver/config/postfix-accounts.cf
             type: File
        - name: opendkim
          hostPath:
            path: /mnt/k8s/yaml/mailserver/config/opendkim
            type: Directory
        - name: data
          persistentVolumeClaim:
            claimName: mailserver-pvc
        - name: tmp-files
          emptyDir: {}
        - name: tls
          secret:
            secretName: <redacted this, can you guess why?>
```

### Relevant Stack Traces

```BASH
# BEGIN

# END
```

**Comment by georglauterbach** [comment]
Will have a look in a few days, I think I saw this as well, but can't remember the solution ATM.

Question: Where did you get the deployment from? We're parts of this taken from the currently open issue

**Comment by georglauterbach** [comment]
The reason _should_ be this line:

```
drop: [ALL]
```

(see the code formatting I did. there :D)

Try running without this, and then add more capabilities until this error is resolved.

**Comment by colindawson** [comment]
My deployment came in part from several places.
I started with the deployment from the documentation, then started tweaking it.

I moved quite a few of the config files out of the configmap.yaml and put them on the file system.  That way I can use the setup.sh file to create mailboxes and things like that.

My service, is using MetalLb, and I've got that set up so that my service passes the client IP properly though to the container.

The bit that you mentioned about for the security context did indeed come from #2094 as I was hoping that they'd solved the issue.  However, that appears not be the case as the permission error is happening inside the container itself.

I've also tried added the following to my volume mounts

```
          - name: data
            mountPath: /var/lib/amavis/tmp
            subPath: amavis/tmp
          - name: data
            mountPath: /var/lib/amavis/db
            subPath: amavis/db
          - name: tls
            mountPath: /etc/ssl/mailserver
            readOnly: true
```

This does solve the permission error, however it replaces it with another problem....

Aug  8 14:18:25 mailserver-777bddfd8d-w299c amavis[1260]: (!!)TROUBLE in pre_loop_hook: TEMPBASE directory is not writable: /var/lib/amavis/tmp at /usr/sbin/amavisd-new line 13180.
Aug  8 14:18:25 mailserver-777bddfd8d-w299c amavis[1260]: (!)_DIE: Suicide () TROUBLE in pre_loop_hook: TEMPBASE directory is not writable: /var/lib/amavis/tmp at /usr/sbin/amavisd-new line 13180.

**Comment by colindawson** [comment]
I tried another variation on the above

```
          - name:  tmp-files
            mountPath: /var/lib/amavis/tmp
            subPath: amavis/tmp
          - name:  tmp-files
            mountPath: /var/lib/amavis/db
            subPath: amavis/db
```

This results in progress, but I'm now getting the following error

Aug  8 14:30:25 mailserver-66cf8cc596-ks56r amavis[525]: (00525-01) (!)connect to /var/run/clamav/clamd.ctl failed, attempt #1: Can't connect to a UNIX socket /var/run/clamav/clamd.ctl: No such file or directory
Aug  8 14:30:26 mailserver-66cf8cc596-ks56r amavis[525]: (00525-01) (!)connect to /var/run/clamav/clamd.ctl failed, attempt #1: Can't connect to a UNIX socket /var/run/clamav/clamd.ctl: No such file or directory
Aug  8 14:30:26 mailserver-66cf8cc596-ks56r amavis[525]: (00525-01) (!)ClamAV-clamd: All attempts (1) failed connecting to /var/run/clamav/clamd.ctl, retrying (2)

**Comment by colindawson** [comment]
Digging into things a bit more I found a freshclam.log

ClamAV update process started at Sun Aug  8 12:00:01 2021
Sun Aug  8 12:00:01 2021 -> WARNING: Your ClamAV installation is OUTDATED!
Sun Aug  8 12:00:01 2021 -> WARNING: Local version: 0.103.2 Recommended version: 0.103.3
Sun Aug  8 12:00:01 2021 -> DON'T PANIC! Read https://www.clamav.net/documents/upgrading-clamav
Sun Aug  8 12:00:01 2021 -> daily database available for update (local version: 26256, remote version: 26257)
Sun Aug  8 12:02:00 2021 -> Testing database: '/var/lib/clamav/tmp.d3365d0888/clamav-80b11e578c5c18f5cba170c3cbb2d081.tmp-daily.cld' ...
Sun Aug  8 12:02:26 2021 -> Database test passed.
Sun Aug  8 12:02:27 2021 -> daily.cld updated (version: 26257, sigs: 1966944, f-level: 90, builder: raynman)
Sun Aug  8 12:02:28 2021 -> main.cvd database is up-to-date (version: 61, sigs: 6607162, f-level: 90, builder: sigmgr)
Sun Aug  8 12:02:28 2021 -> bytecode.cvd database is up-to-date (version: 333, sigs: 92, f-level: 63, builder: awillia2)
Sun Aug  8 12:02:28 2021 -> Clamd successfully notified about the update.

I've seen that others have had trouble with a mismatch between clam-av and clamd versions.
Not sure if this is anything, but thought I'd mention it for information.

**Comment by colindawson** [comment]
hmm, I'm not sure exactly why, but.
it's working.

I think the key was setting the amavis tmp and db folders to use tmp-files rather than a persisted volume.
I discovered it started working as I'm also taking a look into getting the server to do SPF checks, haven't figured that out yet (and if off topic for this issue)    but in checking the logs, I found that I'm now seeing the following in the logs

Aug  8 14:56:14 mailserver-66cf8cc596-4h8dq amavis[1004]: (01004-02) Passed CLEAN {RelayedInbound}, [209.85.161.50]:38697 [209.85.161.50] <c.j.dawson76@gmail.com> -> <me@cjdawson.com>, Queue-ID: 8AEDEDB283, Message-ID: <cda0d471-3c28-5c85-3dcc-ad9719518523@gmail.com>, mail_id: NXhaomQe_7G6, Hits: 1.632, size: 2705, queued_as: 9A90FDB285, 2986 ms
